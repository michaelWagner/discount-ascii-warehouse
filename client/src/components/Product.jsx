import React from 'react';
import Cart from './Cart';

class Product extends React.Component {

  formatDateInRelativeTime(date) {
    let result = '';
    const timeAddedInMsec = Date.parse(date) / 1000;
    const currentTimeInMsec = (Date.parse(new Date())) / 1000;
    const timeRemainingInMsec = currentTimeInMsec - timeAddedInMsec;
    const unit = {
      'week':   Math.floor( timeRemainingInMsec / 604800 % 52 ),
      'day':    Math.floor( timeRemainingInMsec / 86400 % 7 ),
      'hour':   Math.floor( timeRemainingInMsec / 3600 % 24 ),
      'minute': Math.floor( timeRemainingInMsec / 60 % 60 ),
      'second': Math.floor( timeRemainingInMsec % 60 )
    };

    if (unit['week'] < 1) {
      for ( var key in unit ) {
        // Pluralize units for readability.
        if ( unit[key] > 1 ) {
          result += unit[key] + ' ' + key + 's ';
        }
        if ( unit[key] === 1 ) {
          result += unit[key] + ' ' + key + ' ';
        }
      }
      result += 'ago.'
    } else {
      result = date;
    }

    return result;
  }

  handleClick(event) {
    event.preventDefault();
    this.props.toggleCart(this.props.product);
  }

  styleFace(size) {
    return {'fontSize': size + 'px'};
  }

  render() {
    return (
      <tr key={this.props.product.id} onClick={this.handleClick.bind(this)}>
        <td>{this.props.product.id}</td>
        <td>{this.props.formatPrice(this.props.product.price)}</td>
        <td>{this.props.product.size}</td>
        <td style={this.styleFace(this.props.product.size)}>{this.props.product.face}</td>
        <td>{this.formatDateInRelativeTime(this.props.product.date)}</td>
        {this.props.parentComponent === "Cart" ? <td> x </td> : <td> + </td>}
      </tr>
    );
  }
}
export default Product;
