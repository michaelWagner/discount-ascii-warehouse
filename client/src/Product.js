import React from 'react';
import ProductGrid from './ProductGrid';
import ProductAttribute from './ProductAttribute';

class Product extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (<tr>
      {this.props.product.map((productAttribute, idx) => (
        <ProductAttribute productAttribute={productAttribute} />
      ))}
    </tr>);
  }
}
export default Product;
