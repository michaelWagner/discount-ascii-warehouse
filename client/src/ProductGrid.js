
import React from 'react';
import Client from './Client';
import Product from './Product';

class ProductGrid extends React.Component {
  constructor() {
    super();
    this.state = { products: [] };
  }

  componentDidMount(query) {
    Client.loadProducts(query, (products) => {
      for (var i = 0; i < products.length; i++) {
        if (products[i] === null || products[i] === undefined || products[i] === '') {
          delete products[i];
        }
      }
      console.log(products)
      this.setState({products: products});
    });
  }

  addProductToCart = (product) => {
    const newProduct = product;
  }

  render() {
    const { products } = this.state;

    const productRows = products.map((product, idx) => (
      <tr key={idx} onClick={this.addProductToCart}>
        <td>{JSON.parse(product).id.toString()}</td>
        <td>{JSON.parse(product).size.toString()}</td>
        <td>{JSON.parse(product).price.toString()}</td>
        <td>{JSON.parse(product).face.toString()}</td>
        <td>{JSON.parse(product).date.toString()}</td>
      </tr>
    ));
    console.log(productRows)

    return (
      <div id='product-grid'>
        <table className='ui selectable structured large table'>
          <thead>
            <tr>
              <th>ID</th>
              <th>Size</th>
              <th>Price</th>
              <th>Face</th>
              <th>Date</th>
            </tr>
          </thead>
          <tbody>{productRows}</tbody>
        </table>
      </div>
    );
  }
}

export default ProductGrid;
