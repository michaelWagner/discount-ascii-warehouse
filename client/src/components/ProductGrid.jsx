
import React from 'react';
import Client from './Client';
import Product from './Product';
import Advertisement from './Advertisement';

class ProductGrid extends React.Component {
  constructor() {
    super();
    this.state = { products: [] };
  }

  componentDidMount() {
    this.loadProducts('limit=50');
  }

  loadProducts(query) {
    Client.loadProducts(query, (products) => {
      for (var i = 0; i < products.length; i++) {
        if (products[i] === null || products[i] === undefined || products[i] === '') {
          // Remove trailing newlines or empty nodes.
          delete products[i];
        } else {
          products[i] = JSON.parse(products[i]);
        }
      }
      this.setState({products: products});
    });
  }

  render() {
    const { products } = this.state;

    const productRows = products.map((product, idx) => (
      ((idx % 20 === 0 && idx !== 0)
        ? <Advertisement key={idx}/>
        : <Product key={idx} product={product} onClick={this.addProductToCart} />
      )
    ));

    return (
      <div id='product-grid'>
        <table className='ui selectable structured large table'>
          <thead>
            <tr className="product-attribute-header">
              <th>ID</th>
              <th>Price</th>
              <th>Size</th>
              <th>Face</th>
              <th>Date</th>
            </tr>
          </thead>
          <tbody>{productRows}</tbody>
        </table>
      </div>
    );
  }
}

export default ProductGrid;
