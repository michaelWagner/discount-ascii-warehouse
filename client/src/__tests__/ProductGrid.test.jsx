import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Product from '../components/Product';
import ProductGrid from '../components/ProductGrid';

describe(ProductGrid, () => {

  const mockHasProductGridLoaded = jest.fn(true);
  
  const component = shallow(
    <ProductGrid hasProductGridLoaded={mockHasProductGridLoaded}/>
  );

  it('renders and matches our snapshot', () => {
    const component = renderer.create(
      <ProductGrid hasProductGridLoaded={mockHasProductGridLoaded}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has not loaded all products yet', () => {
    expect(component.state('allProductsLoaded')).toBeFalsy();
  });
});
