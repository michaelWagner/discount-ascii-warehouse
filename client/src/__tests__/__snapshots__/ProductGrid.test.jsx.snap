exports[`class ProductGrid extends /* istanbul ignore next */_react2.default.Component {
  constructor(props) {
    super(props);

    this.state = {
      products: [],
      nextProducts: [],
      isProductGridVisible: false,
      allProductsLoaded: false,
      sortBy: null,
      descending: false,
      page: 0,
      numProductsVisible: 0 };


    this.defaultProps = {
      numProductsPerPage: 25,
      componentName: 'ProductGrid' };

  }

  componentDidMount() {
    // Initially load first batch of products.
    const query =
    'limit=' + this.defaultProps.numProductsPerPage +
    '&skip=' + this.state.page * this.defaultProps.numProductsPerPage;


    this.loadProducts(query, products => {
      const numProductsVisible = products.length;
      const initialProducts = products;
      const nextPage = this.state.page + 1;

      this.setState({
        products: initialProducts,
        numProductsVisible: numProductsVisible,
        isProductGridVisible: true,
        page: nextPage },

      () => {
        // Pre-fetch next batch of products and cache them.
        const nextQuery =
        'limit=' + this.defaultProps.numProductsPerPage +
        '&skip=' + this.state.page * this.defaultProps.numProductsPerPage;

        this.loadProducts(nextQuery, nextProducts => {
          this.setState({ nextProducts: nextProducts });
        });
        // Show ProductGrid now that it has loaded.
        this.props.hasProductGridLoaded();
      });

    });
  }

  hasMoreProducts() {
    return this.state.allProductsLoaded ? false : true;
  }

  loadProducts(query, callback) {
    // Creates an array of products in JSON format.
    /* istanbul ignore next */_Client2.default.loadProducts(query, products => {
      for (var i = 0; i < products.length; i++) {
        if (products[i] === null || products[i] === undefined || products[i] === '') {
          // Remove trailing newlines or empty nodes.
          products.splice(i, 1);
        } else {
          products[i] = JSON.parse(products[i]);
        }
      }

      callback(products);
    });
  }

  loadNextProducts() {
    // Render the page with the previously cached nextProducts and update state.
    const newProducts = this.state.products.concat(this.state.nextProducts);
    const numProductsVisible = newProducts.length;
    const nextPage = this.state.page + 1;
    const query =
    'limit=' + this.defaultProps.numProductsPerPage +
    '&skip=' + numProductsVisible;


    // Update products with cached next page.
    this.setState({
      products: newProducts,
      numProductsVisible: numProductsVisible,
      page: nextPage },

    this.loadProducts(query, nextProducts => {
      if (nextProducts.length < this.defaultProps.numProductsPerPage) {
        this.setState({
          allProductsLoaded: true,
          nextProducts: nextProducts });

      } else {
        // Pre-fetch next products and cache.
        this.setState({ nextProducts: nextProducts });
      }
    }));

  }

  _sort(event) {
    // Make column lowercase and then
    // remove the sort arrows before checking column name.
    var column = event.target.innerHTML.toLowerCase().split(" ")[0];
    // Don't allow sorting on the "Add To Cart" column.
    if (column !== "add") {
      var products = this.state.products.slice();
      var descending = this.state.sortBy === column && !this.state.descending;

      products.sort((a, b) => {
        switch (column) {
          case 'id':
            // Sorts by index which prefixes the id string.
            // Separate index from id, then convert index to int before sorting.
            let idA = parseInt(a[column].split('-')[0], 10);
            let idB = parseInt(b[column].split('-')[0], 10);
            return descending ?
            idB - idA :
            idA - idB;
          case 'price':
            return descending ?
            b[column] - a[column] :
            a[column] - b[column];
          case 'product':
            return descending ?
            b['size'] - a['size'] :
            a['size'] - b['size'];
          case 'size':
            return descending ?
            b[column] - a[column] :
            a[column] - b[column];
          case 'date':
            return descending ?
            new Date(b[column]) - new Date(a[column]) :
            new Date(a[column]) - new Date(b[column]);
          default:
            throw new Error('Column is not valid');}

      });

      this.setState({
        products: products,
        sortBy: column,
        descending: descending });

    }
  }

  render() {/* istanbul ignore next */var _state =






    this.state; /* istanbul ignore next */const products = _state.products,isProductGridVisible = _state.isProductGridVisible,sortBy = _state.sortBy,descending = _state.descending,allProductsLoaded = _state.allProductsLoaded;

    const productRows = products.map(function (product, idx) {
      // Show an Advertisement every 20 Products.
      if (idx % 20 === 0 && idx !== 0) {
        // If there is an Advertisement, render both Advertisement and Product.
        return (
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tbody', /* istanbul ignore next */{ key: idx, __source: { fileName: _jsxFileName, lineNumber: 174 }, __self: this },
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_Advertisement2.default, /* istanbul ignore next */{ key: idx,
              generateRandomId: this.props.generateRandomId,
              componentName: this.defaultProps.componentName, __source: { fileName: _jsxFileName, lineNumber: 175 }, __self: this }),
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_Product2.default, /* istanbul ignore next */{ key: product.id,
              product: product,
              formatPriceFromCentsToDollars: this.props.formatPriceFromCentsToDollars,
              toggleProductInCart: this.props.toggleProductInCart, __source: { fileName: _jsxFileName, lineNumber: 178 }, __self: this })));

      } else {
        // If there is no Advertisement, just render Product.
        return (/* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tbody', /* istanbul ignore next */{ key: idx, __source: { fileName: _jsxFileName, lineNumber: 185 }, __self: this },
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_Product2.default, /* istanbul ignore next */{ key: product.id,
              product: product,
              formatPriceFromCentsToDollars: this.props.formatPriceFromCentsToDollars,
              toggleProductInCart: this.props.toggleProductInCart, __source: { fileName: _jsxFileName, lineNumber: 186 }, __self: this })));

      }
    }.bind(this));

    const headers = ["Product", "Date", "Id", "Price", "Size", "Add To Cart"];

    const productHeaders = headers.map((title, idx) =>
    sortBy === title.toLowerCase() ?
    /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'th', /* istanbul ignore next */{ key: idx, onClick: this._sort.bind(this), __source: { fileName: _jsxFileName, lineNumber: 198 }, __self: this }, title += descending ? ' \u2191' : ' \u2193') :
    /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'th', /* istanbul ignore next */{ key: idx, onClick: this._sort.bind(this), __source: { fileName: _jsxFileName, lineNumber: 199 }, __self: this }, title));



    let grid = null;

    if (isProductGridVisible) {
      grid =
      /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ id: 'product-grid-wrapper', __source: { fileName: _jsxFileName, lineNumber: 207 }, __self: this },
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_reactInfiniteScroller2.default, /* istanbul ignore next */{ className: 'product-grid',
            pageStart: 0,
            loadMore: this.loadNextProducts.bind(this),
            hasMore: !allProductsLoaded, __source: { fileName: _jsxFileName, lineNumber: 208 }, __self: this },
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'table', /* istanbul ignore next */{ className: 'product-grid-table', __source: { fileName: _jsxFileName, lineNumber: 212 }, __self: this },
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'thead', /* istanbul ignore next */{ __source: { fileName: _jsxFileName, lineNumber: 213 }, __self: this },
              /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tr', /* istanbul ignore next */{ className: 'product-attribute-header', __source: { fileName: _jsxFileName, lineNumber: 214 }, __self: this },
                productHeaders)),


            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tfoot', /* istanbul ignore next */{ className: allProductsLoaded ? "product-grid-footer" : "hidden-product-grid-footer", __source: { fileName: _jsxFileName, lineNumber: 218 }, __self: this },
              /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tr', /* istanbul ignore next */{ className: 'table-footer', __source: { fileName: _jsxFileName, lineNumber: 219 }, __self: this },
                /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'td', /* istanbul ignore next */{ colSpan: '6', __source: { fileName: _jsxFileName, lineNumber: 220 }, __self: this },
                  /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ __source: { fileName: _jsxFileName, lineNumber: 221 }, __self: this },
                    "~ end of catalogue ~")))),




            productRows)));




    } else {
      grid = /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ id: 'product-grid', __source: { fileName: _jsxFileName, lineNumber: 233 }, __self: this });
    }

    return grid;
  }} renders and matches our snapshot 1`] = `
<div
  id="product-grid" />
`;

exports[`class ProductGrid extends /* istanbul ignore next */_react2.default.Component {
  constructor(props) {
    super(props);

    this.state = {
      products: [],
      nextProducts: [],
      isProductGridVisible: false,
      allProductsLoaded: false,
      sortBy: null,
      descending: false,
      page: 0,
      numProductsVisible: 0 };


    this.defaultProps = {
      numProductsPerPage: 25,
      componentName: 'ProductGrid' };

  }

  componentDidMount() {
    // Initially load first batch of products.
    const query =
    'limit=' + this.defaultProps.numProductsPerPage +
    '&skip=' + this.state.page * this.defaultProps.numProductsPerPage;


    this.loadProducts(query, products => {
      const numProductsVisible = products.length;
      const initialProducts = products;
      const nextPage = this.state.page + 1;

      this.setState({
        products: initialProducts,
        numProductsVisible: numProductsVisible,
        isProductGridVisible: true,
        page: nextPage },

      () => {
        // Pre-fetch next batch of products and cache them.
        const nextQuery =
        'limit=' + this.defaultProps.numProductsPerPage +
        '&skip=' + this.state.page * this.defaultProps.numProductsPerPage;

        this.loadProducts(nextQuery, nextProducts => {
          this.setState({ nextProducts: nextProducts });
        });
        // Show ProductGrid now that it has loaded.
        this.props.hasProductGridLoaded();
      });

    });
  }

  hasMoreProducts() {
    return this.state.allProductsLoaded ? false : true;
  }

  loadProducts(query, callback) {
    // Creates an array of products in JSON format.
    /* istanbul ignore next */_Client2.default.loadProducts(query, products => {
      for (var i = 0; i < products.length; i++) {
        if (products[i] === null || products[i] === undefined || products[i] === '') {
          // Remove trailing newlines or empty nodes.
          products.splice(i, 1);
        } else {
          products[i] = JSON.parse(products[i]);
        }
      }

      callback(products);
    });
  }

  loadNextProducts() {
    // Render the page with the previously cached nextProducts and update state.
    const newProducts = this.state.products.concat(this.state.nextProducts);
    const numProductsVisible = newProducts.length;
    const nextPage = this.state.page + 1;
    const query =
    'limit=' + this.defaultProps.numProductsPerPage +
    '&skip=' + numProductsVisible;


    // Update products with cached next page.
    this.setState({
      products: newProducts,
      numProductsVisible: numProductsVisible,
      page: nextPage },

    this.loadProducts(query, nextProducts => {
      if (nextProducts.length < this.defaultProps.numProductsPerPage) {
        this.setState({
          allProductsLoaded: true,
          nextProducts: nextProducts });

      } else {
        // Pre-fetch next products and cache.
        this.setState({ nextProducts: nextProducts });
      }
    }));

  }

  _sort(event) {
    // Make column lowercase and then
    // remove the sort arrows before checking column name.
    var column = event.target.innerHTML.toLowerCase().split(" ")[0];
    // Don't allow sorting on the "Add To Cart" column.
    if (column !== "add") {
      var products = this.state.products.slice();
      var descending = this.state.sortBy === column && !this.state.descending;

      products.sort((a, b) => {
        switch (column) {
          case 'id':
            // Sorts by index which prefixes the id string.
            // Separate index from id, then convert index to int before sorting.
            let idA = parseInt(a[column].split('-')[0], 10);
            let idB = parseInt(b[column].split('-')[0], 10);
            return descending ?
            idB - idA :
            idA - idB;
          case 'price':
            return descending ?
            b[column] - a[column] :
            a[column] - b[column];
          case 'product':
            return descending ?
            b['size'] - a['size'] :
            a['size'] - b['size'];
          case 'size':
            return descending ?
            b[column] - a[column] :
            a[column] - b[column];
          case 'date':
            return descending ?
            new Date(b[column]) - new Date(a[column]) :
            new Date(a[column]) - new Date(b[column]);
          default:
            throw new Error('Column is not valid');}

      });

      this.setState({
        products: products,
        sortBy: column,
        descending: descending });

    }
  }

  render() {/* istanbul ignore next */var _state =






    this.state; /* istanbul ignore next */const products = _state.products,isProductGridVisible = _state.isProductGridVisible,sortBy = _state.sortBy,descending = _state.descending,allProductsLoaded = _state.allProductsLoaded;

    const productRows = products.map(function (product, idx) {
      // Show an Advertisement every 20 Products.
      if (idx % 20 === 0 && idx !== 0) {
        // If there is an Advertisement, render both Advertisement and Product.
        return (
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tbody', /* istanbul ignore next */{ key: idx, __source: { fileName: _jsxFileName, lineNumber: 174 }, __self: this },
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_Advertisement2.default, /* istanbul ignore next */{ key: idx,
              generateRandomId: this.props.generateRandomId,
              componentName: this.defaultProps.componentName, __source: { fileName: _jsxFileName, lineNumber: 175 }, __self: this }),
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_Product2.default, /* istanbul ignore next */{ key: product.id,
              product: product,
              formatPriceFromCentsToDollars: this.props.formatPriceFromCentsToDollars,
              toggleProductInCart: this.props.toggleProductInCart, __source: { fileName: _jsxFileName, lineNumber: 178 }, __self: this })));

      } else {
        // If there is no Advertisement, just render Product.
        return (/* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tbody', /* istanbul ignore next */{ key: idx, __source: { fileName: _jsxFileName, lineNumber: 185 }, __self: this },
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_Product2.default, /* istanbul ignore next */{ key: product.id,
              product: product,
              formatPriceFromCentsToDollars: this.props.formatPriceFromCentsToDollars,
              toggleProductInCart: this.props.toggleProductInCart, __source: { fileName: _jsxFileName, lineNumber: 186 }, __self: this })));

      }
    }.bind(this));

    const headers = ["Product", "Date", "Id", "Price", "Size", "Add To Cart"];

    const productHeaders = headers.map((title, idx) =>
    sortBy === title.toLowerCase() ?
    /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'th', /* istanbul ignore next */{ key: idx, onClick: this._sort.bind(this), __source: { fileName: _jsxFileName, lineNumber: 198 }, __self: this }, title += descending ? ' ↑' : ' ↓') :
    /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'th', /* istanbul ignore next */{ key: idx, onClick: this._sort.bind(this), __source: { fileName: _jsxFileName, lineNumber: 199 }, __self: this }, title));



    let grid = null;

    if (isProductGridVisible) {
      grid =
      /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ id: 'product-grid-wrapper', __source: { fileName: _jsxFileName, lineNumber: 207 }, __self: this },
        /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */_reactInfiniteScroller2.default, /* istanbul ignore next */{ className: 'product-grid',
            pageStart: 0,
            loadMore: this.loadNextProducts.bind(this),
            hasMore: !allProductsLoaded, __source: { fileName: _jsxFileName, lineNumber: 208 }, __self: this },
          /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'table', /* istanbul ignore next */{ className: 'product-grid-table', __source: { fileName: _jsxFileName, lineNumber: 212 }, __self: this },
            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'thead', /* istanbul ignore next */{ __source: { fileName: _jsxFileName, lineNumber: 213 }, __self: this },
              /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tr', /* istanbul ignore next */{ className: 'product-attribute-header', __source: { fileName: _jsxFileName, lineNumber: 214 }, __self: this },
                productHeaders)),


            /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tfoot', /* istanbul ignore next */{ className: allProductsLoaded ? "product-grid-footer" : "hidden-product-grid-footer", __source: { fileName: _jsxFileName, lineNumber: 218 }, __self: this },
              /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'tr', /* istanbul ignore next */{ className: 'table-footer', __source: { fileName: _jsxFileName, lineNumber: 219 }, __self: this },
                /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'td', /* istanbul ignore next */{ colSpan: '6', __source: { fileName: _jsxFileName, lineNumber: 220 }, __self: this },
                  /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ __source: { fileName: _jsxFileName, lineNumber: 221 }, __self: this },
                    "~ end of catalogue ~")))),




            productRows)));




    } else {
      grid = /* istanbul ignore next */_react2.default.createElement( /* istanbul ignore next */'div', /* istanbul ignore next */{ id: 'product-grid', __source: { fileName: _jsxFileName, lineNumber: 233 }, __self: this });
    }

    return grid;
  }} renders and matches our snapshot 1`] = `
<div
  id="product-grid" />
`;
